From 40df27b76e8952f846ad7369aa49350ab728811b Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Fri, 11 Oct 2024 18:21:42 +0200
Subject: [PATCH 4/4] arm64: dts: add description for solidrun lx2160a cex6
 evalution board

Add description for the Solid-Run internal CEX Type 6 module and
evaluation board.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 arch/arm64/boot/dts/freescale/Makefile        |   1 +
 .../dts/freescale/fsl-lx2160a-cex6-evb.dts    | 234 ++++++++++++++
 .../boot/dts/freescale/fsl-lx2160a-cex6.dtsi  | 294 ++++++++++++++++++
 3 files changed, 529 insertions(+)
 create mode 100644 arch/arm64/boot/dts/freescale/fsl-lx2160a-cex6-evb.dts
 create mode 100644 arch/arm64/boot/dts/freescale/fsl-lx2160a-cex6.dtsi

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 9a7319c6b4e3..810ba23bf1b2 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -39,6 +39,7 @@ dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-ls2081a-rdb.dtb
 dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-ls2080a-simu.dtb
 dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-ls2088a-qds.dtb
 dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-ls2088a-rdb.dtb
+dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-lx2160a-cex6-evb.dtb
 dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-lx2160a-clearfog-cx.dtb
 dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-lx2160a-honeycomb.dtb
 dtb-$(CONFIG_ARCH_LAYERSCAPE) += fsl-lx2160a-qds.dtb
diff --git a/arch/arm64/boot/dts/freescale/fsl-lx2160a-cex6-evb.dts b/arch/arm64/boot/dts/freescale/fsl-lx2160a-cex6-evb.dts
new file mode 100644
index 000000000000..55a4c81023a5
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/fsl-lx2160a-cex6-evb.dts
@@ -0,0 +1,234 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+//
+// Device Tree file for LX2160A-CEx6 Evaluation Board
+//
+// Copyright 2024 SolidRun Ltd.
+
+/dts-v1/;
+
+#include "fsl-lx2160a.dtsi"
+#include "fsl-lx2160a-cex6.dtsi"
+
+/ {
+	model = "SolidRun LX2160A CEX Type 6 Evaluation Board";
+	compatible = "solidrun,lx2160a-cex6-evb", "solidrun,lx2160a-cex6", "fsl,lx2160a";
+
+	aliases {
+		crypto = &crypto;
+		gpio0 = &gpio0;
+		gpio1 = &gpio1;
+		gpio2 = &gpio2;
+		gpio3 = &gpio3;
+		gpio4 = &expander0;
+		gpio5 = &expander1;
+		gpio6 = &expander2;
+		i2c0 = &i2c0;
+		i2c1 = &i2c2;
+		i2c2 = &i2c3;
+		i2c3 = &i2c_eeprom;
+		i2c4 = &i2c_fan;
+		i2c5 = &i2c_reg;
+		i2c6 = &i2c_mon;
+		i2c7 = &i2c_rtc;
+		i2c8 = &i2c_gpio;
+		i2c9 = &b2b_i2c2;
+		i2c10 = &b2b_i2c3;
+		mmc0 = &esdhc0;
+		mmc1 = &esdhc1;
+		rtc0 = &com_rtc;
+		serial0 = &uart0;
+		serial1 = &uart1;
+		serial2 = &uart2;
+		serial3 = &uart3;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	reg_usb_vbus0: regulator-usb-vbus0 {
+		/*
+		 * Power for USB Port 0:
+		 * - enable gpio: EC2_RX_DV / GPIO4[23]
+		 * - over-current alert gpio (output from regulator): IRQ10 / GPIO3[10]
+		 */
+		compatible = "regulator-fixed";
+		regulator-name = "vbus0";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpios = <&gpio3 23 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		/* dwc3 driver does not toggle vbus supply */
+		regulator-always-on;
+	};
+
+	reg_usb_vbus1: regulator-usb-vbus1 {
+		/*
+		 * Power for USB Port 1:
+		 * - enable gpio: EC2_TX_EN / GPIO4[16]
+		 * - over-current alert gpio (output from regulator): IRQ11 / GPIO3[11]
+		 */
+		compatible = "regulator-fixed";
+		regulator-name = "vbus1";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpios = <&gpio3 16 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		/* dwc3 driver does not toggle vbus supply */
+		regulator-always-on;
+	};
+
+	sfp1: sfp-1 { // J67
+		compatible = "sff,sfp";
+		i2c-bus = <&i2c3>;
+		mod-def0-gpios = <&gpio2 3 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	sfp2: sfp-2 { // J49
+		compatible = "sff,sfp";
+		i2c-bus = <&i2c2>;
+		mod-def0-gpios = <&gpio2 4 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+
+	sfp3: sfp-3 { // J69
+		compatible = "sff,sfp";
+		i2c-bus = <&b2b_i2c2>;
+		mod-def0-gpios = <&gpio2 5 GPIO_ACTIVE_LOW>;
+		maximum-power-milliwatt = <2000>;
+	};
+};
+
+&b2b_i2c3 {
+	/* PCA9655 U10532 @ 20 */
+	expander1: gpio@20 {
+		compatible = "onnn,pca9655";
+		reg = <0x20>;
+		gpio-controller;
+		gpio-line-names = "GPI0", "GPI1", "GPI2", "GPI3",
+				  "IRQ4_N", "IRQ5_N", "USB0_ID", "IRQ0",
+				  "GPO0", "GPO1", "GPO2", "GPO3",
+				  "WAKE_N2", "WAKE_N1", "USB1_ID", "EN_VBUS1";
+	};
+};
+
+&dpmac4 {
+	sfp = <&sfp2>;
+	managed = "in-band-status";
+	phys = <&serdes_1 6>;
+	status = "okay";
+};
+
+&dpmac5 {
+	sfp = <&sfp1>;
+	managed = "in-band-status";
+	phys = <&serdes_1 5>;
+	status = "okay";
+};
+
+&dpmac6 {
+	sfp = <&sfp3>;
+	managed = "in-band-status";
+	phys = <&serdes_1 4>;
+	status = "okay";
+};
+
+&dspi2 {
+	w25q32fvzpig0: flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		m25p,fast-read;
+		spi-max-frequency = <50000000>;
+		reg = <0>;
+		spi-rx-bus-width = <1>;
+		spi-tx-bus-width = <1>;
+	};
+
+	w25q32fvzpig1: flash@1 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		m25p,fast-read;
+		spi-max-frequency = <50000000>;
+		reg = <1>;
+		spi-rx-bus-width = <1>;
+		spi-tx-bus-width = <1>;
+	};
+
+	w25q32fvzpig2: flash@2 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		m25p,fast-read;
+		spi-max-frequency = <50000000>;
+		reg = <2>;
+		spi-rx-bus-width = <1>;
+		spi-tx-bus-width = <1>;
+	};
+
+	w25q32fvzpig3: dflash@3 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		m25p,fast-read;
+		spi-max-frequency = <50000000>;
+		reg = <3>;
+		spi-rx-bus-width = <1>;
+		spi-tx-bus-width = <1>;
+	};
+};
+
+&fspi {
+	mt35xu512aba1: flash@1 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		m25p,fast-read;
+		spi-max-frequency = <50000000>;
+		reg = <1>;
+		spi-rx-bus-width = <8>;
+		spi-tx-bus-width = <8>;
+	};
+};
+
+&i2c0 {
+	/* PCA9655 U10541 @ 21 */
+	expander2: gpio@21 {
+		compatible = "onnn,pca9655";
+		reg = <0x21>;
+		gpio-controller;
+		gpio-line-names = "TYPE_N0", "TYPE_N1", "TYPE_N2", "TYPE_N3",
+				  "THERM_WRN_B_3V3", "IO0_5", "IO0_6", "IO0_7",
+				  "IO1_0", "IO1_1", "IO1_2", "IO1_3",
+				  "IO1_4", "IO1_5", "IO1_6", "IO1_7";
+	};
+};
+
+&pcs_mdio4 {
+	status = "okay";
+};
+
+&pcs_mdio5 {
+	status = "okay";
+};
+
+&pcs_mdio6 {
+	status = "okay";
+};
+
+&serdes_1 {
+	/* disable driver for lanes A-D */
+	fsl,unmanaged-lanes = <0x0f>;
+};
+
+&usb0 {
+	vbus-supply = <&reg_usb_vbus0>;
+	status = "okay";
+};
+
+&usb1 {
+	vbus-supply = <&reg_usb_vbus1>;
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/freescale/fsl-lx2160a-cex6.dtsi b/arch/arm64/boot/dts/freescale/fsl-lx2160a-cex6.dtsi
new file mode 100644
index 000000000000..17a8047e25f4
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/fsl-lx2160a-cex6.dtsi
@@ -0,0 +1,294 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+//
+// Device Tree file for LX2160A-CEx6
+//
+// Copyright 2024 SolidRun Ltd.
+
+&crypto {
+	status = "okay";
+};
+
+&dpmac3 {
+	phys = <&serdes_1 7>;
+	phy-names = "serdes";
+	phy-handle = <&ethernet_phy1>;
+	phy-connection-type = "10gbase-r";
+	status = "okay";
+};
+
+&dpmac17 {
+	phy-handle = <&ethernet_phy0>;
+	phy-connection-type = "rgmii-id";
+	status = "okay";
+};
+
+&emdio1 {
+	status = "okay";
+
+	ethernet_phy0: ethernet-phy@1 {
+		reg = <1>;
+		qca,smarteee-tw-us-1g = <24>;
+	};
+
+	/* AQR113 PHY */
+	ethernet_phy1: ethernet-phy@8 {
+		compatible = "ethernet-phy-ieee802.3-c45";
+		reg = <0x8>;
+		interrupt-parent = <&gpio2>;
+		interrupts = <8 IRQ_TYPE_EDGE_FALLING>; // IRQ08 / GPIO3_DAT08
+		reset-gpios = <&gpio0 23 GPIO_ACTIVE_LOW>; // IIC6_SCL / GPIO1_DAT23
+	};
+};
+
+&dspi2 {
+	status = "okay";
+};
+
+&esdhc0 {
+	no-1-8-v;
+	status = "okay";
+};
+
+&esdhc1 {
+	bus-width = <8>;
+	mmc-hs200-1_8v;
+	mmc-hs400-1_8v;
+	status = "okay";
+};
+
+&fspi {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		m25p,fast-read;
+		spi-max-frequency = <50000000>;
+		/* The following setting enables 1-1-8 (CMD-ADDR-DATA) mode */
+		spi-rx-bus-width = <8>;
+		spi-tx-bus-width = <1>;
+	};
+};
+
+&gpio2 {
+	/*
+	 * AMC6821 THERM signal uses open-drain logic and can be driven
+	 * by either the host (force full-speed) or the chip
+	 * (thermal warning). Firmware drives it low during reset.
+	 *
+	 * Release the signal here for autonomous operation.
+	 * TODO: Add support to fan-controller driver.
+	 */
+	fan-full-speed-hog {
+		gpio-hog;
+		gpios = <2 (GPIO_ACTIVE_LOW | GPIO_OPEN_DRAIN)>;
+		input;
+		line-name = "fan-full-speed";
+	};
+};
+
+&i2c0 {
+	status = "okay";
+
+	i2c-switch@77 {
+		compatible = "nxp,pca9547";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x77>;
+
+		i2c_eeprom: i2c@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0>;
+
+			eeprom@50 {
+				compatible = "atmel,24c512";
+				reg = <0x50>;
+			};
+
+			eeprom@51 {
+				compatible = "atmel,spd";
+				reg = <0x51>;
+			};
+
+			eeprom@53 {
+				compatible = "atmel,spd";
+				reg = <0x53>;
+			};
+
+			eeprom@57 {
+				compatible = "atmel,24c02";
+				reg = <0x57>;
+			};
+		};
+
+		i2c_fan: i2c@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <1>;
+
+			fan-temperature-ctrlr@18 {
+				compatible = "ti,amc6821";
+				reg = <0x18>;
+				cooling-min-state = <0>;
+				cooling-max-state = <9>;
+				#cooling-cells = <2>;
+			};
+
+			/* fan-controller@2f (emc2301) */
+
+			/* clock-controller@6a */
+		};
+
+		i2c_reg: i2c@2 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <2>;
+
+			regulator@5c {
+				compatible = "lltc,ltc3882";
+				reg = <0x5c>;
+
+				regulators {
+					/* both outputs are connected, supplying soc core @ 0.8V */
+					vout0 {
+						regulator-name = "VDD_PH1";
+						regulator-always-on;
+					};
+
+					vout1 {
+						regulator-name = "VDD_PH2";
+						regulator-always-on;
+					};
+				};
+			};
+		};
+
+		i2c_mon: i2c@3 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <3>;
+
+			current-mon@40 { //GVDD 1.2V
+				compatible = "ti,ina219";
+				reg = <0x40>;
+				shunt-resistor = <1000>;
+			};
+
+			current-mon@44 { //SD_SVDD 0.9V
+				compatible = "ti,ina219";
+				reg = <0x44>;
+				shunt-resistor = <1000>;
+			};
+
+			current-mon@42 { //SD_OVDD 1.8V
+				compatible = "ti,ina219";
+				reg = <0x42>;
+				shunt-resistor = <1000>;
+			};
+
+			current-mon@45 { //OVDD 1.8V
+				compatible = "ti,ina219";
+				reg = <0x45>;
+				shunt-resistor = <1000>;
+			};
+
+			current-mon@43 { //0.7V AQR113 PHY
+				compatible = "ti,ina219";
+				reg = <0x43>;
+				shunt-resistor = <1000>;
+			};
+
+			current-mon@46 { //5V
+				compatible = "ti,ina219";
+				reg = <0x46>;
+				shunt-resistor = <1000>;
+			};
+
+			current-mon@4d { //3.3V
+				compatible = "ti,ina219";
+				reg = <0x4D>;
+				shunt-resistor = <1000>;
+			};
+
+			temp-mon@4c { // TEMP_DIODE1/2 (U29/U30 - assembly option)
+				compatible = "nxp,sa56004";
+				reg = <0x4c>;
+				#thermal-sensor-cells = <1>;
+			};
+
+			temp-mon@48 { // temp sensor Q11 (bottom side, edge of pcb, near USD connector)
+				compatible = "nxp,sa56004";
+				reg = <0x48>;
+				#thermal-sensor-cells = <1>;
+			};
+		};
+
+		i2c_rtc: i2c@4 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <4>;
+
+			com_rtc: rtc@68 {
+				compatible = "dallas,ds1374";
+				reg = <0x68>;
+				interrupt-parent = <&gpio2>;
+				interrupts = <9 IRQ_TYPE_EDGE_FALLING>; // IRQ2_RTC_N --> IRQ09 / GPIO3_DAT09
+			};
+		};
+
+		i2c_gpio: i2c@5 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <5>;
+
+			/* assembly option */
+			expander0: gpio@41 {
+				compatible = "nxp,pca9536";
+				reg = <0x41>;
+				gpio-controller;
+				gpio-line-names = "USB0_ID", "USB0_VBUS", "USB1_ID", "USB1_VBUS";
+			};
+		};
+
+		b2b_i2c2: i2c@6 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <6>;
+		};
+
+		b2b_i2c3: i2c@7 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <7>;
+		};
+	};
+};
+
+&i2c2 {
+	status = "okay";
+};
+
+&i2c3 {
+	status = "okay";
+};
+
+&pcs_mdio3 {
+	status = "okay";
+};
+
+&uart0 {
+	status = "okay";
+};
+
+&uart1 {
+	status = "okay";
+};
+
+&uart2 {
+	status = "okay";
+};
+
+&uart3 {
+	status = "okay";
+};
-- 
2.43.0

